woocommerce/woocommerce
# Implement server-side remote error logging

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:





Closes https://github.com/woocommerce/woocommerce/issues/49287.

This PR implements server-side (PHP) remote error logging for WooCommerce. This feature is currently disabled by default.

When the feature flag is enabled and meets the criteria, it will capture all Woocomerce-related fatal errors and log them asynchronously to the server. Developer can also use `log()` method to log custom errors.

Some of the key features of this PR are:

- `register_shutdown_function` is used to capture fatal errors (This is how we're currently capturing fatal errors in WooCommerce for status report).
- Sanitize and trace the error message before sending it to the server.
  - ` /var/www/html/wp-content/plugins/woocommerce/includes/class-wc-remote-logger.php on line 123` will be sanitized to: `**/woocommerce/includes/class-wc-remote-logger.php on line 123`
- Set `non-blocking` to `true` in `wp_safe_remote_post` to make the request non-blocking.
- Implement throttling to control the rate of error logging and prevent system overload.
- Filter out errors originating from third parties by dropping error events that exclusively contain third-party stack frames
- Introduce `woocommerce_remote_logger_formatted_log_data` filter to allow other plugins to modify the log data before sending it to the server.
- Introduce `woocommerce_remote_logging_is_third_party_error` filter to allow other plugins to determine if an error is third-party or not. I think this is useful for sending fatal logs to other Woo plugins.

#### Example log entry:

We log the following data to the server:

- host: The host of the site
- message: The error message
- trace: The error trace
- php_version: The PHP version
- wc_version: The WooCommerce version
- wp_version: The WordPress version
- store_id: The store ID

![Screenshot 2024-07-18 at 12 09 32](https://github.com/user-attachments/assets/da79b5c6-bfdf-4cd7-8b3a-f15301264454)

#### Relevant links for reference:

- Logstash API: fbhepr%2Skers%2Sjcpbz%2Schoyvp.ncv%2Serfg%2Sjcpbz%2Qwfba%2Qraqcbvagf%2Spynff.jcpbz%2Qwfba%2Qncv%2Qybtfgnfu%2Qraqcbvag.cuc%3Se%3Qnr27n451-og
- WPCOM WooCommerce Logstash handler: fbhepr%2Skers%2Sjcpbz%2Sjc%2Qpbagrag%2Szh%2Qcyhtvaf%2Sjbbpbzzrepr%2Svapyhqrf%2Spynff%2Qjp%2Qybtfgnfu%2Qunaqyre.cuc%3Se%3Q8p4146o2%2396-og




### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

This PR is only testable for WooCommerce core team members since it requires access to the logstash server.


1. Create a fresh install of WooCommerce and `Code Snippet` plugin
2. Add the following code snippet to your site to enable remote logging feature:

```php
update_option( 'woocommerce_feature_remote_logging_enabled', 'yes' );

add_filter(
	'option_woocommerce_allow_tracking',
	function () {
		return 'yes';
	}
);

add_filter(
	'option_woocommerce_remote_variant_assignment',
	function () {
		return 5;
	}
);
```

3. Trigger a fatal error in WooCommerce by adding the following code to [`load_rest_api()`](https://github.com/woocommerce/woocommerce/blob/ac834f55f75d4e51955bad43115939faa38ec9da/plugins/woocommerce/includes/class-woocommerce.php#L535):

```php
throw new Exception( 'Test fatal error in woocommerce' );
```

4. Refresh the page and check the fatal error is displayed on the page
5. Reverse the changes made in step 3
6. Trigger a fatal error in a third-party plugin by adding the following code to the `load_classes` function in the `wp-content/plugins/code-snippets/php/admin.php` file


```php
		throw new \Exception( 'Code Snippets admin menus are not loaded' );
```

7. Refresh the page and check the fatal error is displayed on the page
8. Check the log server page for the log entry (See test plan in D155228-code to see how to check the log entry)
9. Confirm that woocommerce fatal errors are logged and third-party fatal errors are not logged
10. Please test around the feature and review the code changes 



### Changelog entry







<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
